### PR TITLE
Contact Points and Contact Manifold

### PR DESCRIPTION
**Pull request by fallenatlas**
# Description

- Add Contact Manifold
- Compute contact points using the Sutherland Hodgman algorithm.
- Updated collisions sample to show the resulting info on the contact manifold.

(yup, that's all)

To support this:

- Added a Plane to geom
- Added the concept of faces and edges to box shape
- Added utils to geom, which contains functions that don't directly handle intersections.

I'll gladly take any suggestions so Box.hpp doesn't look so ugly.

Resources:
https://research.ncl.ac.uk/game/mastersdegree/gametechnologies/previousinformation/physics5collisionmanifolds/2017%20Tutorial%205%20-%20Collision%20Manifolds.pdf
https://research.ncl.ac.uk/game/mastersdegree/gametechnologies/physicstutorials/4collisiondetection/Physics%20-%20Collision%20Detection.pdf
Also used the structures to hold the date in Box2D/Avian for reference.

## Checklist

- [x] Self-review changes.
- [x] Evaluate impact on the documentation.
- [x] Add entry to the changelog's unreleased section.

### CODE REVIEW

**Review by joaomanita — APPROVED** [review]

File: core/include/cubos/core/geom/intersections.hpp [hunk 1]
```diff
@@ -33,4 +34,24 @@ namespace cubos::core::geom
     CUBOS_CORE_API bool intersects(const Box& box1, const glm::mat4& localToWorld1, const Box& box2,
                                    const glm::mat4& localToWorld2, Intersection& intersect);
 
+    /// @brief Performs the Sutherland-Hodgman Clipping algorithm.
+    /// @param inputPolygon The polygon to perform the clipping on.
+    /// @param numClipPlanes Number of cliping planes.
+    /// @param clipPlanes Clipping planes
+    /// @param removeNotClipToPlane Whether to remove the points if they're outside the plane.
+    /// @return The polygon resulting from the clipping.
+    CUBOS_CORE_API std::list<glm::vec3> sutherlandHodgmanClipping(const std::list<glm::vec3>& inputPolygon,
```
Comment (RiscadoA) [inline core/include/cubos/core/geom/intersections.hpp:43]: Any reason for an `std::list` btw? I think a `std::vector` would probably perform better. (Same for all other uses of it in this PR)

**Review by RiscadoA — APPROVED** [review]
Incredible work, ty very much for working on this! Also ran the sample on my PC, very cool to see the results :eyes:
Performance wise I'm sure we can improve this a lot further but it's not worth it to waste time on that right now. The only thing I think you should change on that part is replacing `std::list` by `std::vector` - unless you've a reason for sticking to the former ofc.

Also sorry for the review delay once again!

**Comment by RiscadoA** [comment]
Also it would be cool to link those resources in the code itself so that it doesn't get forgotten

**Review by Dacops — APPROVED** [review]
post resolving the pending conversations; lgtm

File: core/include/cubos/core/geom/box.hpp [hunk 2]
```diff
@@ -51,6 +53,221 @@ namespace cubos::core::geom
             corners[7] = {halfSize.x, -halfSize.y, -halfSize.z};
         }
 
+        /// @brief Computes the enclosing faces of the corner.
+        /// @param vertexIndex Index of the corner, according to the returned by corners.
+        /// @param faces Array to store the three faces in.
```
[the remark this answers is not in the document]
Comment (fallenatlas) [reply core/include/cubos/core/geom/box.hpp:58]: This one stores the indexes of the faces that have that vertex, according to the array returned by the function that gives the faces. Will improve the documentation a little bit, bc it's not super clear

File: core/include/cubos/core/geom/box.hpp [hunk 3]
```diff
@@ -51,6 +53,221 @@ namespace cubos::core::geom
...
+            else
+            {
+                faces[0] = 0;
+                faces[1] = 3;
+                faces[2] = 4;
+            }
+        }
+
+        /// @brief Computes the edges of a box.
+        /// @param edges Array to store the twelve edges in.
```
[the remark this answers is not in the document]
Comment (fallenatlas) [reply core/include/cubos/core/geom/box.hpp:112]: An edge is simply a pair of corner indices yeah, the corner indexes that make that "aresta" of the box.

File: core/include/cubos/core/geom/intersections.hpp (hunk 1 above)
Comment (fallenatlas) [reply core/include/cubos/core/geom/intersections.hpp:43]: I initially just followed along the example code, and I remember thinking about it too, but I don't remember if I saw anything or I just forgot to change. I will review and change if I see it's ok.

**Pull request merged**